cli/cache/cache_db.rs

```diff
@@ -0,0 +1,468 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::error::AnyError;
+use deno_core::parking_lot::Mutex;
+use deno_core::parking_lot::MutexGuard;
+use deno_runtime::deno_webstorage::rusqlite;
+use deno_runtime::deno_webstorage::rusqlite::Connection;
+use deno_runtime::deno_webstorage::rusqlite::OptionalExtension;
+use deno_runtime::deno_webstorage::rusqlite::Params;
+use once_cell::sync::OnceCell;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::sync::Barrier;
+
+/// What should the cache should do on failure?
+#[allow(unused)]
+pub enum CacheFailure {
+  /// Create an in-memory cache that is not persistent.
+  InMemory,
+  /// Create a blackhole cache that ignores writes and returns empty reads.
+  Blackhole,
+  /// Return errors.
+  Error,
+}
+
+/// Configuration SQL and other parameters for a [`CacheDB`].
+pub struct CacheDBConfiguration {
+  /// SQL to run for a new database.
+  pub table_initializer: &'static str,
+  /// SQL to run when the version from [`crate::version::deno()`] changes.
+  pub on_version_change: &'static str,
+  /// Prepared statements to pre-heat while initializing the database.
+  pub preheat_queries: &'static [&'static str],
+  /// What the cache should do on failure.
+  pub on_failure: CacheFailure,
+}
+
+impl CacheDBConfiguration {
+  fn create_combined_sql(&self) -> String {
+    format!(
+      "
+      PRAGMA journal_mode=OFF;
+      PRAGMA synchronous=NORMAL;
+      PRAGMA temp_store=memory;
+      PRAGMA page_size=4096;
+      PRAGMA mmap_size=6000000;
+      PRAGMA optimize;
+
+      CREATE TABLE IF NOT EXISTS info (
+        key TEXT PRIMARY KEY,
+        value TEXT NOT NULL
+      );
+
+      {}
+    ",
+      self.table_initializer
+    )
+  }
+}
+
+enum ConnectionState {
+  Connected(Connection),
+  Blackhole,
+  Error(Arc<AnyError>),
+}
+
+/// A cache database that eagerly initializes itself off-thread, preventing initialization operations
+/// from blocking the main thread.
+#[derive(Clone)]
+pub struct CacheDB {
+  // TODO(mmastrac): We can probably simplify our thread-safe implementation here
+  conn: Arc<Mutex<OnceCell<ConnectionState>>>,
+  path: Option<PathBuf>,
+  config: &'static CacheDBConfiguration,
+  version: &'static str,
+}
+
+impl Drop for CacheDB {
+  fn drop(&mut self) {
+    let arc = std::mem::take(&mut self.conn);
+    if let Ok(inner) = Arc::try_unwrap(arc) {
+      // Hand off SQLite connection to another thread to do the surprisingly expensive cleanup
```

Comment:
Do you know what kind of cleanup it's doing? I think this increases the chance of the cleanup not happening when the process exits. I'd assume that's probably ok or does that increase the chance of corruption?

Reply:
Good question. I confirmed that using `spawn_blocking` will correctly wait for this shutdown to happen (added a 10s timeout to drop and the process waited for that to complete).